### PR TITLE
Deleting in batches

### PR DESCRIPTION
**Pull request by paterczm**
https://github.com/lightblue-platform/lightblue-mongo/issues/268

Work in progress. Feedback welcome, but do not merge yet.

### CODE REVIEW

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java [hunk 1]
```diff
@@ -35,14 +47,103 @@
...
+                }
+
+                if (docsToDelete.size() == batchSize || !cursor.hasNext()) {
+                    // batch built or run out of documents
+
+                    // TODO: write concern
+                    BulkWriteOperation bw = collection.initializeUnorderedBulkOperation();
+
+                    List<Object> _ids = docsToDelete.stream().map(d -> d._id).collect(Collectors.toList());
+                    bw.find(new BasicDBObject("_id", new BasicDBObject("$in", _ids))).remove();
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java]: No need for bulk ops here. collection.remove(inQuery} should do

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java [hunk 2]
```diff
@@ -35,14 +47,103 @@
...
+            }
+        }
+
+        response.setNumDeleted(numDeleted);
+    }
+
+    private void handleBulkWriteError(List<BulkWriteError> errors, List<DocInfo> docs) {
+        for (BulkWriteError e : errors) {
+            DocInfo doc = docs.get(e.getIndex());
+            // TODO: what error should we use here?
```
Comment (paterczm) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java]: Please comment on this, I'm not sure which error code to use here. Maybe we need a new one?

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java [hunk 3]
```diff
@@ -35,14 +46,109 @@
...
+
+    private Translator translator;
+
+    public BasicDocDeleter(Translator translator) {
+        super();
+        this.translator = translator;
+    }
+
+    private final class DocInfo {
+        public DocInfo(DocCtx docCtx, Object _id) {
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java]: Not used, remove

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java [hunk 4]
```diff
@@ -35,14 +46,109 @@
...
+
+        response.setNumDeleted(numDeleted);
+    }
+
+    private void handleBulkWriteError(List<BulkWriteError> errors, List<DocInfo> docs) {
+        for (BulkWriteError e : errors) {
+            DocInfo doc = docs.get(e.getIndex());
+            // TODO: what error should we use here?
+            doc.docCtx.addError(Error.get("remove", MongoCrudConstants.ERR_SAVE_ERROR, e.getMessage()));
+        }
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java]: ERR_DELETE_ERROR, add if there isn't one already

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java (hunk 2 above)
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java]: remove() doesn't throw any meaningful exceptions, a generic DELETE exception should do.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoCRUDController.java [hunk 5]
```diff
@@ -336,7 +336,7 @@ public CRUDDeleteResponse delete(CRUDOperationContext ctx,
                 LOGGER.debug("Translated query {}", mongoQuery);
                 DB db = dbResolver.get((MongoDataStore) md.getDataStore());
                 DBCollection coll = db.getCollection(((MongoDataStore) md.getDataStore()).getCollectionName());
-                DocDeleter deleter = new IterateDeleter(translator);
+                DocDeleter deleter = new BasicDocDeleter(translator);
```
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoCRUDController.java]: You can remove IterateDeleter class.

**Comment by paterczm** [comment]
After adding ERR_DELETE_ERROR, https://github.com/lightblue-platform/lightblue-mongo/issues/106 should be also covered.

**Pull request merged**